cql3/statements/select_statement.cc

```diff
@@ -831,6 +830,32 @@ indexed_table_select_statement::prepare(database& db,
     const auto& im = index_opt->metadata();
     sstring index_table_name = im.name() + "_index";
     schema_ptr view_schema = db.find_schema(schema->ks_name(), index_table_name);
+    std::optional<global_index_prepare_data> global_index_data;
+
+    if (!im.local()) {
+        // Fill in the global_index_prepare_data structure
+        const column_definition& token_cdef = *view_schema->clustering_key_columns().begin();
+
+        const bool is_token_v2_index = token_cdef.type == long_type;
+        const auto& partition_key_restriction_expression = restrictions->get_partition_key_restrictions()->expression;
+        // Global posting list partition slice according to the token restriction is only supported
+        // with new token_column_computation type (computation name: token_v2)
+        bool is_token_partition_slice = has_token(partition_key_restriction_expression) && is_token_v2_index;
+
+        ::shared_ptr<restrictions::single_column_clustering_key_restrictions> posting_list_clustering_restrictions;
+        if (is_token_partition_slice) {
+            posting_list_clustering_restrictions = ::make_shared<restrictions::single_column_clustering_key_restrictions>(view_schema, false);
+
+            auto token_restriction = ::make_shared<restrictions::single_column_restriction>(token_cdef);
+            token_restriction->expression = expr::replace_token(partition_key_restriction_expression, &token_cdef);
+
+            posting_list_clustering_restrictions->merge_with(token_restriction);
+        }
+
+        global_index_data.emplace(global_index_prepare_data{
```

Comment:
Nitpick: with `.emplace`, why invoke the constructor explicitly?

Reply:
I had issues with clang not supporting it well.

Reply:
Tried those two options:

```cpp
global_index_data.emplace(
    token_cdef, is_token_partition_slice, std::move(posting_list_clustering_restrictions)
);
```

```cpp
global_index_data.emplace({
    token_cdef, is_token_partition_slice, std::move(posting_list_clustering_restrictions)
});
```

And they don't compile under clang 10.0.0. (Didn't debug this extensively, but it seems to be caused by `token_cdef` being reference and `emplace` treating it as not a reference? `candidate template ignored: could not match 'initializer_list<type-parameter-0-0>' against 'column_definition'`)

Reply:
I think it's p0960r3, parenthesized initialization of aggregates. emplace() calls a parenthesized constructor, which gcc 10 accepts due to p0960r3, but clang rejects since it doesn't have it.

Reply:
OK

Reply:
What happens if you just say `global_index_data = global_index_prepare_data{...}`?

Reply:
Then the copy constructor of `global_index_prepare_data` is called and it fails to compile (`object of type 'std::optional<global_index_prepare_data>' cannot be assigned because its copy assignment operator is implicitly deleted
` - probably related to reference inside it?).